src/lib.rs

```diff
@@ -6,10 +6,173 @@
+    #[repr(transparent)]
+    #[derive(Copy, Clone, Debug, Eq, PartialEq)]
+    struct Flags: u16 {
+        /// Clear to send.
+        const CTS = 1 << 0;
+        /// Data set ready.
+        const DSR = 1 << 1;
+        /// Data carrier detect.
+        const DCD = 1 << 2;
+        /// UART busy transmitting data.
+        const BUSY = 1 << 3;
+        /// Receive FIFO is empty.
+        const RXFE = 1 << 4;
+        /// Transmit FIFO is full.
+        const TXFF = 1 << 5;
+        /// Receive FIFO is full.
+        const RXFF = 1 << 6;
+        /// Transmit FIFO is empty.
+        const TXFE = 1 << 7;
+        /// Ring indicator.
+        const RI = 1 << 8;
+    }
+}
+
+bitflags! {
+    /// Flags from the UART Receive Status Register / Error Clear Register.
+    #[repr(transparent)]
+    #[derive(Copy, Clone, Debug, Eq, PartialEq)]
+    struct ReceiveStatus: u16 {
+        /// Framing error.
+        const FE = 1 << 0;
+        /// Parity error.
+        const PE = 1 << 1;
+        /// Break error.
+        const BE = 1 << 2;
+        /// Overrun error.
+        const OE = 1 << 3;
+    }
+}
 
+#[repr(C, align(4))]
+struct Registers {
+    /// Data Register.
+    dr: u16,
+    _reserved0: [u8; 2],
+    /// Receive Status Register.
+    rsr: ReceiveStatus,
+    _reserved1: [u8; 19],
+    /// Flag Register.
+    fr: Flags,
+    _reserved2: [u8; 6],
+    /// IrDA Low-Power Counter Register.
+    ilpr: u8,
+    _reserved3: [u8; 3],
+    /// Integer Baud Rate Register.
+    ibrd: u16,
+    _reserved4: [u8; 2],
+    /// Fractional Baud Rate Register.
+    fbrd: u8,
+    _reserved5: [u8; 3],
+    /// Line Control Register.
+    lcr_h: u8,
+    _reserved6: [u8; 3],
+    /// Control Register.
+    cr: u16,
+    _reserved7: [u8; 3],
+    /// Interrupt FIFO Level Select Register.
+    ifls: u8,
+    _reserved8: [u8; 3],
+    /// Interrupt Mask Set/Clear Register.
+    imsc: u16,
+    _reserved9: [u8; 2],
+    /// Raw Interrupt Status Register.
+    ris: u16,
+    _reserved10: [u8; 2],
+    /// Masked Interrupt Status Register.
+    mis: u16,
+    _reserved11: [u8; 2],
+    /// Interrupt Clear Register.
+    icr: u16,
+    _reserved12: [u8; 2],
+    /// DMA Control Register.
+    dmacr: u8,
+    _reserved13: [u8; 3],
+}
+
+/// Driver for a PL011 UART.
+#[derive(Debug)]
+pub struct Uart {
+    registers: *mut Registers,
 }
 
 impl Uart {
+    /// Constructs a new instance of the UART driver for a PL011 device at the
+    /// given base address.
+    ///
+    /// # Safety
+    ///
+    /// The given base address must point to the MMIO control registers of a
+    /// PL011 device, which must be mapped into the address space of the process
+    /// as device memory and not have any other aliases.
+    pub unsafe fn new(base_address: *mut u32) -> Self {
+        Self {
+            registers: base_address as *mut Registers,
+        }
+    }
+
+    /// TODO: implement init
+    /// TODO: crash context vs non-crash context?
```

Comment:
What does crash context mean in practice?

Reply:
I noticed that the C tf-a seems to split the UART functions (into core and non-core) to avoid memory access in the core functions, so that it can be used in crash context 
https://github.com/ARM-software/arm-trusted-firmware/blob/master/drivers/arm/pl011/aarch64/pl011_console.S#L13

Reply:
Do the crash context ones there need to avoid stack access? If so then that can't be in Rust, it will need to remain in assembly. I suggest we don't worry about that in this driver.

Reply:
I think so, yeah, let me get rid of that TODO